### PR TITLE
map-button

### LINKED ISSUE #685: Feature Request: map mouse button
### Use Case

mod-middle-click to toggle floating

### Prior Art

* dwm [config.def.h](https://git.suckless.org/dwm/file/config.def.h.html#l99)
* i3 `bindsym --whole-window $mod1+button2 floating toggle`

### Options

1. Overload `map-pointer` to allow all commands
2. Create `map-button` to allow mapping button to a command
3. Allow `map` to take input event code

1 is unacceptable as `map-pointer` is specifically for mouse motions.

3 would be simplest for the user. Implementation would also be simpler than 2.

### PR DESCRIPTION
**Pull request by alex-courtis**
resolves #685 

Adds `map-button` and `unmap-button`: executes a command similar to `map`, using an event code instead of a keysym.

Does not feature `[-release|-repeat|-layout index]`.

Silently overrides `map-pointer`. Please advised whether a warning or automatic `unmap-pointer` is required.

First time using zig; pointers and corrections would be gratefully appreciated.

### CODE REVIEW

**Comment by alex-courtis** [comment]
### Testing Performed

```
riverctl map-button normal Super BTN_MIDDLE toggle-float

riverctl map-button normal Super+Control BTN_MIDDLE toggle-fullscreen

riverctl unmap-button normal Super+Control BTN_MIDDLE

riverctl unmap-button normal Super+Control BTN_MIDDLE

riverctl map-button normal Super+Control BTN_MIDDLE
error: not enough arguments

riverctl map-button normal Super+Control BTN_BLEH toggle-fullscreen
error: unknown button BTN_BLEH

riverctl map-button normal Super+Bleh BTN_MIDDLE toggle-fullscreen
error: invalid modifier 'Bleh'

riverctl map-button bleh Super+Control BTN_MIDDLE toggle-fullscreen
error: cannot add/remove mapping to/from non-existant mode 'bleh'

riverctl unmap-button normal Super+Control BTN_MIDDLE toggle-fullscreen
error: too many arguments

riverctl unmap-button normal Super+Control
error: not enough arguments
```

Same mapping with different commands for different modes: correct commands executed


Same mapping for map-button and map-pointer: map-button overrides map-pointer


Ensured button not sent to client on click


zsh completions but not bash/fish

File: river/ButtonMapping.zig [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2020 The River Developers
```
Comment (Leon-Plickat) [inline river/ButtonMapping.zig:3]:
```suggestion
// Copyright 2022 The River Developers
```

File: river/Seat.zig [hunk 2]
```diff
@@ -405,6 +405,23 @@ pub fn handleMapping(
...
 
+/// Handle any user-defined mapping for passed button and modifiers.
+/// Returns true if there was a mapping.
+pub fn handleButtonMapping(
+    self: *Self,
+    event_code: u32,
+    modifiers: wlr.Keyboard.ModifierMask,
+) bool {
+    const modes = &server.config.modes;
+    for (modes.items[self.mode_id].button_mappings.items) |*mapping| {
```
Comment (Leon-Plickat) [inline river/Seat.zig:416]:
```suggestion
    for (modes.items[self.mode_id].button_mappings.items) |mapping| {
```

You only need a pointer if you want to modify the item.

**Comment by Leon-Plickat** [comment]
Seems a bit weird to me to have both `map-buttom` and `map-pointer` TBH. Maybe it would make sense to combine the two?

**Comment by alex-courtis** [comment]
> Seems a bit weird to me to have both `map-buttom` and `map-pointer` TBH. Maybe it would make sense to combine the two?

Sounds good. keyboard vs mouse can be inferred from keycode vs event code.
Looks like that was my initial idea #685 ...

`-release|-repeat` don't really make sense for buttons. Suggestion:
* mapping always occurs on press
* return an error if the user attempts to set those options

`-layout index` should be implementable.

File: river/Seat.zig (hunk 2 above)
Comment (alex-courtis) [reply river/Seat.zig:416]: Thank you!

**Comment by Leon-Plickat** [comment]
> keyboard vs mouse can be inferred from keycode vs event code.

What keyboard? Both `map-pointer` and `map-button` are for the mouse.

**Comment by alex-courtis** [comment]
> > keyboard vs mouse can be inferred from keycode vs event code.
> 
> What keyboard? Both `map-pointer` and `map-button` are for the mouse.

Whoops... I got that backwards. I was thinking of overloading `map` for keyboard and mouse.

Proposal:
`map-pointer mode modifiers button action|command`

**Comment by alex-courtis** [comment]
Superceeded by #692

**Pull request closed without merging**